doc/libfoa.3
.TH "LIBFOA" "3" "2009-03-06" "Nowise Systems" ""
\" This manual page is under GNU general public license (GPL).
\" Copyright (C) 2009-2019 Anders Lövgren
.SH NAME
C library implementing FOA - Fast Object and Array encoding/decoding

.SH SYNOPSIS
\fB#include <libfoa.h>\fR        /* standard API */
.br
\fB#include <libfoa-utils.h>\fR  /* utility API */

.SH DESCRIPTION
.PP
This is a library implementation of the FOA encoding, designed to be an 
alternative to XML for data exchange between i.e. web services. It can also useful
as a general encoding format similar to JSON.
.PP
The memory footprint of the library is small because it will never use more 
memory than required to hold the longest single line of data to decode.
.PP
The library can decode or encoded data using either a memory buffer (char *) 
or an file stream (FILE *). No validation is performed except for some 
rudimentary checks. The library itself is thread safe.

.SH TYPES
This section briefly describes the FOA encoding. Each encoded line contains
either data (could be an array entry or an object member) or an special char. Each 
special char or data item must be followed by a newline (\\n) character. The 
special chars are 
.B '(' 
(begin object),
.B '[' 
(begin array),
.B ']' 
(end array) and
.B ')' 
(end object).
.PP
Each data, array or object (type) can also be named by 'name = type'. Arrays and 
objects can be unlimited nested in each other.
.PP
.B Syntax:
.PP
.br 
item := (name '=' data | data) + '\\n'
.br 
data := text | spec
.br 
name := text - '='
.br 
text := any char
.br 
spec := ([])
.br 
.PP
.B Example:
.PP
This is an example of how an array (named arr) containing two structures
(persons with named members) will be encoded:
.PP
arr = [
.br
(
.br
name = Albert Einstein
.br
born = March 14, 1879
.br
)
.br
(
.br
name = Isaac Newton
.br
born = December 25, 1642
.br
)
.br
]
.PP
.B Escape:
.PP
If escaping is turned on (the default), then special chars will be escaped when encoding
and unescaped when decoding (they are mapped to HTTP encoded equivalents).
This is all transparent to the library user. The escaping can be turned off
thru \fBfoa_set_mode()\fR to improve performance.

.SH OPTIONS
The libfoa structure is opaque and you are not supposed to peek inside it.
.PP
.RS
.nf
.B struct foa_entity
{
	const char *name;      /* entity name (might be NULL) */
	const char *data;      /* entity data (decoded data) */
	int type;              /* entity type (see FOA_TYPE_XXX) */
	unsigned long line;    /* current input line */
};
.RE
.fi
.TP 
.B typedef int (*foa_handler)(const struct foa_entity *ent, void *arg);
Prototype for the callback function set by foa_set_callback() and used by
foa_scan() when decoding data. This function is called both for scanned data
and as an error handler. If this function returns 0 to foa_scan(), then the 
parsing will stop and foa_scan() will return immidiatelly.
.TP
.B foa_init(struct libfoa *foa);
Initilize library usage. Returns 0 if successful and -1 on error.
.TP
.B void foa_cleanup(struct libfoa *foa);
Cleanup after finished. This function should only be called when done with
the library.
.TP
.B int foa_alloc_strategy(struct libfoa *foa, size_t step, size_t max);
Defines the memory allocation strategy. The step variable defines how
aggressive memory is allocated. The variable max sets a maximum
buffer size in bytes (use 0 for unlimited). If max is smaller than current
used buffer size, then the buffer is reallocated. Returns 0 if sucessful 
and -1 on error. 
.TP
.B void foa_set_callback(struct libfoa *foa, foa_handler func, void *arg);
Set a callback function to handle decoded data. The optional arg is
passed along together with each call to func.
.TP 
.B void foa_set_stream(struct libfoa *foa, FILE *file);
Set a file stream for reading and writing.
.TP
.B void foa_set_buffer(struct libfoa *foa, const char *input);
Sets the input buffer for memory reading. You can't set the buffer for writing,
its handled internal.
.TP 
.B void foa_set_mode(struct libfoa *foa, int mode, int enable);
Sets the current mode for escaping of special chars and whether to support
named data. Turning of escaping will improve performance, but you must be
sure that data items (text) will never contain any special chars '([])'. See
\fBMODES\fR section also.
.TP 
.B void foa_get_mode(struct libfoa *foa, int mode, int *enable);
Get current mode for escaping and named data (hashes). The mode argument is
either FOA_MODE_ESCAPE or FOA_MODE_HASHES.
.TP 	
.B int foa_scan(struct libfoa *foa);
Decode the input using the registered callback function for each item in
input. Returns 0 when input data has been successful processed and -1 on
error.
.TP
.B const struct foa_entity * foa_next(struct libfoa *foa);
Decode the input iterative. This is an alternative to using the foa_scan() 
function. Returns the next entity or NULL on error.
.TP
.B const char * foa_write(struct libfoa *foa, int append, int type, 
                          \fBconst char *name, const char *data);\fR
.br			  
Write data to memory buffer or file stream. If append is true, then data is 
appended (only for memory buffer), otherwise the memory buffer is overwritten.
The type is one of FOA_TYPE_XXX. Use NULL as name to write anonymous data. 
Returns the encoded string on success and NULL on error.

.SH MODES
This section describes the contants that can be used as mode argument for
foa_get_mode() and foa_set_mode().
.TP
.B FOA_MODE_ESCAPE
Enable or disable escaping of special chars when encoding or decoding.
Escaping of data is enabled by default.
.TP 
.B FOA_MODE_HASHES
Controls whether named objects, arrays and data entities are allowed, this
is the default.
.TP
.B FOA_MODE_SETERR
Decode errors are always inlined in the entity structure by foa_next()
and foa_scan(). Call foa_set_mode() using FOA_MODE_SETERR as mode argument
to control whether the error message should be written to last error buffer
as well so it can be fetched by calling foa_last_error(), this is the default.

.SH ERRORS
These functions is used to access the last error message.
.TP 
.B const char * foa_last_error(struct libfoa *foa);
Get last error message.
.TP 
.B int foa_error_set(struct libfoa *foa);
Returns true if error is set.
.TP 
.B void foa_reset_error(struct libfoa *foa);
Clear the stored error message.

.SH EXAMPLE
An simple example showing how to decode data from a file stream:
.RS
.nf

#include <stdio.h>
#include <libfoa.h>

int main(int argc, char **argv)
{
    struct libfoa foa;
    const struct foa_entity *ent;
    FILE *fs;
				     
    if(foa_init(&foa) < 0) {
        fprintf(stderr, "failed initilize libfoa: %s\\n", foa_last_error(&foa));
        return 1;
    }
    
    fs = fopen(argv[1], "r");
    if(!fs) {
        fprintf(stderr, "failed open %s for reading\\n", argv[1]);
        return 1;
    }

    foa_set_stream(&foa, fs);
    while((ent = foa_next(&foa))) {
        switch(ent->type) {
        case FOA_TYPE_DATA_ENTITY:
            if(ent->name)
                printf("%s: %s\\n", ent->name, ent->data);
            else
                printf("data: %s\\n", ent->data);
            break;
        default:
            printf("spec: %s\\n", ent->data);
            break;
        }
    }
	  
    fclose(fs);
    foa_cleanup(&foa);
			  
    return 0;
}				  
.RE
.fi

.SH SPECIFICATION
.PP 
The official FOA specification can be read at: 
.br
.B http://it.bmc.uu.se/andlov/proj/libfoa/spec.php

.SH NOTES
Full length examples of encoding/decoding and more documentation of the FOA
format is distributed with the source code.

.SH BUGS
Perhaps, send an bug report to andlov@nowise.se if you find one.

.SH AUTHOR
Anders Lövgren


NEWS
Version 0.6.11 (2019-01-29):

        * Update autoconf files (autoreconf -i)

Version 0.6.10 (2019-01-29):

        * Migrated to GIT.

        * New maintainer address, bug report and homepage URL.

Version 0.6.9 (2013-06-11):

	* Fixed build issues using current autoconf, automake and libtool.
	
	* Converted project to UTF-8 encoding.
	
Version 0.6.8 (2009-05-19):

	* Fixed bug where named objects and arrays got their type set to
	  first character from data. Introduced in version 0.6.7, fixed here.
	  
	* The last error buffer will now contain decode errors (syntax)
	  unless explicit turned off by calling foa_set_mode().

Version 0.6.7 (2009-05-19):

	* Naming objects and arrays where impossible in previous version,
	  fixed now.

Version 0.6.6 (2009-03-15):

	* Updated the documentation explaining some domains of application of FOA.
	
Version 0.6.5 (2009-03-15):

	* Integrated building the examples in the MS VS2003 solution.
	
	* Switching to release mode. Fixed all warnings building with VC7.
	
	* Enable conditional build of examples (pass --enable-examples to configure).
	
Version 0.6.4 (2009-03-13):

	* Added makefile for building examples using MS VS2003 (nmake).
	
Version 0.6.3 (2009-03-11):

	* Fixed DLL-export of some utility functions when building on Windows.
	
	* Added makefile for building examples using Borland C++ 5.5 (Windows).
	
Version 0.6.2 (2009-03-10):

	* Some minor bug fixes, among them adding this NEWS file ;-)

Version 0.6.1 (2009-03-10):

	* Windows port: Added support for building the library using Borland
	  C++ 5.5 (the free command line tools from Inprise).

Version 0.6.0 (2009-03-09):

	* Windows port: Added support for buidling the library using
	  Microsoft Visual Studio 2003 (professional).

	* Lots of fixes regarding export of public API
	
Version 0.5.0 (2009-03-06):
	
	* First release. Includes library + example and test code. Supports
	  Linux/UNIX using GCC.


src/libfoa.c
/* Library implementing FOA - Fast Object and Array encoding.
 * Copyright (C) 2009-2019 Anders Lövgren
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#ifdef HAVE_STDLIB_H
# include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
# include <string.h>
#elif HAVE_STRINGS_H
# include <strings.h>
#endif
#include <errno.h>

#include "libfoa.h"
#include "libfoa-utils.h"
#include "internal.h"

/*
 * Initilize library.
 */
int foa_init(struct libfoa *foa)
{
	memset(foa, 0, sizeof(struct libfoa));
	
	foa->size = FOA_MEMORY_ALLOC_INIT;
	foa->step = FOA_MEMORY_ALLOC_STEP;
	foa->max  = FOA_MEMORY_ALLOC_MAX;
	foa->buff = malloc(foa->size);
	if(!foa->buff) {
		write_errlog(&foa->errmsg, errno, "failed alloc memory");
		return -1;
	}
	foa->escape = FOA_ENABLE_ESCAPE;
	foa->hashes = FOA_ENABLE_HASHES;
	foa->seterr = FOA_ENABLE_SETERR;
	
	return 0;
}

/* 
 * Cleanup after finished.
 */
void foa_cleanup(struct libfoa *foa)
{
	free(foa->buff);
	free(foa->errmsg);
}

/* 
 * Define memory allocation strategy. The step variable defines how
 * aggressive memory is allocated. The max variable sets a maximum
 * buffer size in bytes (use 0 for unlimited). Returns 0 if successful
 * and -1 on error.
 */
int foa_alloc_strategy(struct libfoa *foa, size_t step, size_t max)
{
	foa->step = step;
	foa->max  = max;
	if(foa->max < foa->used && foa->max != 0) {
		foa->buff = realloc(foa->buff, foa->max);
		if(!foa->buff) {
			write_errlog(&foa->errmsg, errno, "failed alloc memory");
			return -1;
		}
		foa->used = foa->max;
	}
	return 0;
}

/* 
 * Use callback function to handle data seen. The optional arg get
 * passed along together with each call to func.
 */
void foa_set_callback(struct libfoa *foa, foa_handler func, void *arg)
{
	foa->func = func;
	foa->arg  = arg;
}

/* 
 * Set file stream for reading (parsing) and writing (encoded result).
 */
void foa_set_stream(struct libfoa *foa, FILE *file)
{
	foa->line = 1;
	foa->used = 0;
	foa->curr = foa->next = NULL;
	foa->file = file;
}

/* 
 * Set input buffer for memory parsing.
 */
void foa_set_buffer(struct libfoa *foa, const char *input)
{
	foa->line = 1;
	foa->used = 0;
	foa->file = NULL;
	foa->curr = input;
	foa->next = strchr(foa->curr, '\n');	
}

/* 
 * Set current mode.
 */
void foa_set_mode(struct libfoa *foa, int mode, int enable)
{
	switch(mode) {
	case FOA_MODE_ESCAPE:
		foa->escape = enable == 0 ? 0 : 1;
		break;
	case FOA_MODE_HASHES:
		foa->hashes = enable == 0 ? 0 : 1;
		break;
	}
}
	
/* 
 * Get current mode.
 */
void foa_get_mode(struct libfoa *foa, int mode, int *enable)
{
	switch(mode) {
	case FOA_MODE_ESCAPE:
		*enable = foa->escape;
		break;
	case FOA_MODE_HASHES:
		*enable = foa->hashes;
		break;
	}
}

/* 
 * Process the input buffer or stream. This calls the registered callback 
 * function for each type (data or special character) seen.
 */
int foa_scan(struct libfoa *foa)
{
	if(!foa->func) {
		logerr(&foa->errmsg, 0, "no callback function defined");
		return -1;
	}
	if(!foa->curr && !foa->file) {
		logerr(&foa->errmsg, 0, "no input stream or memory buffer is set");
		return -1;
	}
	
	if(foa->file) {
		while(!feof(foa->file)) {
			if(read_stream(foa) < 0) {
				return feof(foa->file) ? 0 : -1;
			}
			if(!foa->func(&foa->entity, foa->arg)) {
				return -1;
			}
		}
	} else {
		while(foa->curr) {
			if(read_memory(foa) < 0) {
				return -1;
			}
			if(!foa->func(&foa->entity, foa->arg)) {
				return -1;
			}
		}
	}
	return 0;
}

/* 
 * Process the input buffer iterative. This is an alternative to
 * using the foa_scan() function.
 */
const struct foa_entity * foa_next(struct libfoa *foa)
{
	if(foa->file) {
		if(feof(foa->file)) {
			return NULL;
		}
		if(read_stream(foa) < 0) {
			return NULL;
		}
	} else if(foa->curr) {
		if(read_memory(foa) < 0) {
			return NULL;
		}
	} else {
		if(!foa->curr && foa->line != 1) {
			return NULL;  /* end of buffer */
		}
		logerr(&foa->errmsg, 0, "input stream or memory is unset");
		return NULL;
	}
	return &foa->entity;
}

/* 
 * Clear stored error message.
 */
void foa_reset_error(struct libfoa *foa)
{
	free(foa->errmsg);
	foa->errmsg = NULL;
}

/* 
 * Specialized for writing anonymous string data, that is, array of 
 * const char * where last element must be NULL. This function may 
 * cause the maximum alloc limit to be reached for the memory buffer 
 * if used in append mode.
 */
const char * foa_write_array(struct libfoa *foa, const char **data, int append)
{
	const char *res;
	
	res = foa_write(foa, append, FOA_TYPE_START_ARRAY, NULL, NULL);
	if(!res) 
		return NULL;
	
	while(*data) {
		res = foa_write(foa, append, FOA_TYPE_DATA_ENTITY, NULL, *(data++));
		if(!res)
			return NULL;
	}
	res = foa_write(foa, append, FOA_TYPE_END_ARRAY, NULL, NULL);
	if(!res)
		return NULL;
	
	return foa->buff;
}
